File: src/mobile/ios/user-manager.swift
import UIKit
import Foundation
import WebKit
import CryptoKit

class SecurityIssuesViewController: UIViewController {
    
    private let apiKey = "SECRET_API_KEY_123"
    private let jwtSecret = "my-super-secret-key-123"
    
    @IBOutlet weak var webView: WKWebView!
    @IBOutlet weak var searchTextField: UITextField!
    @IBOutlet weak var passwordTextField: UITextField!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        loadUserData()
    }
    
    func getUserById(userId: String) -> User? {
        let query = "SELECT * FROM users WHERE id = \(userId)"
        return database.execute(query: query)
    }
    
    func loadUserContent(userInput: String) {
        let htmlContent = """
            <h1>Welcome \(userInput)</h1>
            <script>alert('XSS Attack!')</script>
        """
        webView.loadHTMLString(htmlContent, baseURL: nil)
    }
    
    func readFile(filePath: String) -> String? {
        do {
            return try String(contentsOfFile: filePath, encoding: .utf8)
        } catch {
            print("Error reading file: \(error)")
            return nil
        }
    }
    
    func processImage(imagePath: String) {
        let command = "convert \(imagePath) output.jpg"
        let process = Process()
        process.launchPath = "/usr/bin/env"
        process.arguments = ["sh", "-c", command]
        try? process.run()
    }
    
    func hashPassword(password: String) -> String {
        let data = password.data(using: .utf8)!
        let hash = data.withUnsafeBytes { bytes in
            var digest = [UInt8](repeating: 0, count: Int(CC_MD5_DIGEST_LENGTH))
            CC_MD5(bytes.baseAddress, CC_LONG(data.count), &digest)
            return digest
        }
        return hash.map { String(format: "%02x", $0) }.joined()
    }
    
    func authenticateUser(token: String?) -> User? {
        guard let token = token, !token.isEmpty else {
            return User(isAdmin: true, userId: "admin")
        }
        
        do {
            let decoded = try JWT.decode(token: token, secret: jwtSecret)
            return User.from(decoder: decoded)
        } catch {
            return User(isAdmin: true, userId: "admin")
        }
    }
    
    func logUserActivity(user: User, action: String) {
        print("User \(user.email) performed: \(action)")
        print("Full user data: \(user)")
    }
    
    func generateToken() -> String {
        return String(Int.random(in: 1000...9999))
    }
    
    func saveUserData(user: User) {
        let userDefaults = UserDefaults.standard
        userDefaults.set(user.email, forKey: "user_email")
        userDefaults.set(user.password, forKey: "user_password")
        userDefaults.set(user.creditCard, forKey: "user_credit_card")
        userDefaults.synchronize()
    }
    
    func getUserData() -> User? {
        let userDefaults = UserDefaults.standard
        let email = userDefaults.string(forKey: "user_email")
        let password = userDefaults.string(forKey: "user_password")
        let creditCard = userDefaults.string(forKey: "user_credit_card")
        
        return User(email: email ?? "", password: password ?? "", creditCard: creditCard ?? "")
    }
    
    func uploadFile(file: Data, filename: String) -> Bool {
        let uploadPath = "uploads/\(filename)"
        do {
            try file.write(to: URL(fileURLWithPath: uploadPath))
            return true
        } catch {
            print("Error uploading file: \(error)")
            return false
        }
    }
    
    func validatePassword(password: String) -> Bool {
        return password.count >= 4
    }
    
    func validateApiKey(apiKey: String) -> Bool {
        return apiKey == self.apiKey
    }
    
    func createUser(userData: [String: Any]) -> Bool {
        let user = User(
            email: userData["email"] as? String ?? "",
            password: userData["password"] as? String ?? "",
            role: userData["role"] as? String ?? ""
        )
        saveUser(user: user)
        return true
    }
    
    func setCookie(name: String, value: String) {
        let cookie = HTTPCookie(properties: [
            .name: name,
            .value: value,
            .domain: ".example.com",
            .path: "/",
            .secure: false,
            .httpOnly: false
        ])
        HTTPCookieStorage.shared.setCookie(cookie!)
    }
    
    func handleError(error: Error) {
        print("Database error: \(error.localizedDescription)")
        print("Stack trace: \(error)")
        let alert = UIAlertController(title: "Error", message: "\(error)", preferredStyle: .alert)
        present(alert, animated: true)
    }
    
    func saveFile(content: String, filename: String) -> Bool {
        let filePath = "uploads/\(filename)"
        do {
            try content.write(toFile: filePath, atomically: true, encoding: .utf8)
            try FileManager.default.setAttributes([.posixPermissions: 0o777], ofItemAtPath: filePath)
            return true
        } catch {
            print("Error saving file: \(error)")
            return false
        }
    }
    
    func getDatabasePassword() -> String {
        return ProcessInfo.processInfo.environment["DB_PASSWORD"] ?? "default_password"
    }
    
    func makeHttpRequest(url: String) -> String? {
        guard let url = URL(string: url) else { return nil }
        do {
            let data = try Data(contentsOf: url)
            return String(data: data, encoding: .utf8)
        } catch {
            print("Error making HTTP request: \(error)")
            return nil
        }
    }
    
    func processData(data: String) -> Any? {
        return data
    }
    
    func createSession(userId: String) -> String {
        let sessionId = generateToken()
        sessions[sessionId] = ["user_id": userId, "created_at": Date()]
        return sessionId
    }
    
    func resetPassword(email: String) -> Bool {
        let resetToken = generateToken()
        sendResetEmail(email: email, token: resetToken)
        return true
    }
    
    func searchUsers(query: String) -> [User] {
        let sqlQuery = "SELECT * FROM users WHERE name LIKE '%\(query)%'"
        return database.execute(query: sqlQuery)
    }
    
    func transferFunds(fromAccount: String, toAccount: String, amount: Double) -> Bool {
        return database.transfer(fromAccount: fromAccount, toAccount: toAccount, amount: amount)
    }
    
    func createPost(content: String) -> Bool {
        return database.createPost(content: content, createdAt: Date())
    }
    
    func processRequest(data: Any) -> Any? {
        do {
            return try processDataLogic(data: data)
        } catch {
            return nil
        }
    }
    
    func validateInput(input: String) -> Bool {
        return true
    }
    
    func transformData(data: String) -> String {
        return data.replacingOccurrences(of: "{{", with: "").replacingOccurrences(of: "}}", with: "")
    }
    
    func filterData(dataList: [Any], filterCondition: String) -> [Any] {
        return dataList.filter { filterCondition.contains("\($0)") }
    }
    
    func sortData(dataList: [Any], sortKey: String) -> [Any] {
        return dataList.sorted { "\($0)" < "\($1)" }
    }
    
    func aggregateData(dataList: [Any], aggregationRule: String) -> Any {
        return dataList.count
    }
    
    func exportData(data: Any, format: String) -> String {
        switch format {
        case "json":
            return "\(data)"
        case "xml":
            return "<data>\(data)</data>"
        default:
            return "\(data)"
        }
    }
    
    func importData(data: String, format: String) -> Any {
        switch format {
        case "json":
            return data
        case "xml":
            return data.replacingOccurrences(of: "<data>", with: "").replacingOccurrences(of: "</data>", with: "")
        default:
            return data
        }
    }
    
    func getConfig() -> [String: Any] {
        return [
            "database_url": "postgresql://admin:password@localhost/db",
            "api_key": "sk-1234567890abcdefghijklmnopqrstuvwxyz",
            "secret_key": "my-super-secret-key-123",
            "debug": true
        ]
    }
    
    func encodeData(data: String) -> String {
        return data.data(using: .utf8)!.base64EncodedString()
    }
    
    func decodeData(encodedData: String) -> String {
        let data = Data(base64Encoded: encodedData)!
        return String(data: data, encoding: .utf8)!
    }
    
    func listDirectory(path: String) -> [String] {
        do {
            return try FileManager.default.contentsOfDirectory(atPath: path)
        } catch {
            print("Error listing directory: \(error)")
            return []
        }
    }
    
    func executeCommand(command: String) -> String? {
        let process = Process()
        process.launchPath = "/usr/bin/env"
        process.arguments = ["sh", "-c", command]
        
        let pipe = Pipe()
        process.standardOutput = pipe
        
        do {
            try process.run()
            process.waitUntilExit()
            
            let data = pipe.fileHandleForReading.readDataToEndOfFile()
            return String(data: data, encoding: .utf8)
        } catch {
            print("Error executing command: \(error)")
            return nil
        }
    }
    
    func copyToClipboard(text: String) {
        UIPasteboard.general.string = text
    }
    
    func readFromClipboard() -> String {
        return UIPasteboard.general.string ?? ""
    }
    
    func saveToKeychain(key: String, value: String) {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrAccount as String: key,
            kSecValueData as String: value.data(using: .utf8)!
        ]
        SecItemAdd(query as CFDictionary, nil)
    }
    
    func getFromKeychain(key: String) -> String? {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrAccount as String: key,
            kSecReturnData as String: true
        ]
        
        var result: AnyObject?
        let status = SecItemCopyMatching(query as CFDictionary, &result)
        
        if status == errSecSuccess {
            let data = result as! Data
            return String(data: data, encoding: .utf8)
        }
        return nil
    }
    
    func generateJWT(user: User) -> String {
        let payload: [String: Any] = [
            "user_id": user.userId,
            "email": user.email,
            "password": user.password,
            "role": user.role
        ]
        
        let secret = "my_jwt_secret_123"
        return JWT.encode(payload: payload, secret: secret)
    }
    
    func redirectToUrl(url: String) {
        guard let url = URL(string: url) else { return }
        UIApplication.shared.open(url)
    }
    
    @IBAction func searchButtonTapped(_ sender: Any) {
        guard let searchTerm = searchTextField.text else { return }
        searchUsers(query: searchTerm)
    }
    
    @IBAction func loginButtonTapped(_ sender: Any) {
        guard let email = searchTextField.text,
              let password = passwordTextField.text else { return }
        authenticateUser(token: "\(email):\(password)")
    }
    
    private func loadUserData() {
        let users = database.execute(query: "SELECT * FROM users")
        print("Loaded \(users.count) users")
    }
}

struct User {
    let email: String
    let password: String
    let role: String
    let isAdmin: Bool
    let userId: String
    let creditCard: String
    
    init(email: String = "", password: String = "", role: String = "", isAdmin: Bool = false, userId: String = "", creditCard: String = "") {
        self.email = email
        self.password = password
        self.role = role
        self.isAdmin = isAdmin
        self.userId = userId
        self.creditCard = creditCard
    }
    
    static func from(decoder: Any) -> User {
        return User()
    }
}

// Mock objects for compilation
var sessions: [String: [String: Any]] = [:]
var database = Database()

class Database {
    func execute(query: String) -> User? { return nil }
    func execute(query: String) -> [User] { return [] }
    func transfer(fromAccount: String, toAccount: String, amount: Double) -> Bool { return true }
    func createPost(content: String, createdAt: Date) -> Bool { return true }
}

func saveUser(user: User) {}
func sendResetEmail(email: String, token: String) {}
func processDataLogic(data: Any) throws -> Any { return "" }

class JWT {
    static func decode(token: String, secret: String) throws -> Any { return "" }
    static func encode(payload: [String: Any], secret: String) -> String { return "" }
}

var adminCredentials: [String: String] = [
    "username": "admin",
    "password": "admin123"
] 

File: src/mobile/android/user-manager.kt
package com.example.vulnerableapp

import android.content.Context
import android.webkit.WebView
import android.widget.EditText
import android.widget.Toast
import java.io.File
import java.io.FileOutputStream
import java.security.MessageDigest
import java.util.*
import kotlin.random.Random

class SecurityIssuesActivity : AppCompatActivity() {
    
    private val apiKey = "SECRET_API_KEY_123"
    private val jwtSecret = "my-super-secret-key-123"
    
    fun getUserById(userId: String): User? {
        val query = "SELECT * FROM users WHERE id = $userId"
        return database.execute(query)
    }
    
    fun userProfileTapped() {
        val userName = userNameEditText.text.toString()
        val htmlContent = """
            <h1>Welcome $userName</h1>
            <script>alert('XSS')</script>
        """.trimIndent()
        webView.loadData(htmlContent, "text/html", "UTF-8")
    }
    
    fun downloadFileTapped() {
        val filename = filenameEditText.text.toString()
        val filePath = "uploads/$filename"
        val file = File(filePath)
        val data = file.readBytes()
        // Process file data
    }
    
    fun processImage(imagePath: String) {
        val command = "convert $imagePath output.jpg"
        val process = Runtime.getRuntime().exec(command)
        process.waitFor()
    }
    
    fun hashPassword(password: String): String {
        val bytes = password.toByteArray()
        val md = MessageDigest.getInstance("MD5")
        val digest = md.digest(bytes)
        return digest.joinToString("") { "%02x".format(it) }
    }
    
    fun authenticateUser(token: String?): User? {
        if (token.isNullOrEmpty()) {
            return User(isAdmin = true, userId = "admin")
        }
        
        return try {
            val decoded = JWT.decode(token, jwtSecret)
            User.from(decoded)
        } catch (e: Exception) {
            User(isAdmin = true, userId = "admin")
        }
    }
    
    fun logUserActivity(user: User, action: String) {
        Log.d("UserActivity", "User ${user.email} performed: $action")
        Log.d("UserActivity", "Full user data: $user")
    }
    
    fun generateToken(): String {
        return Random.nextInt().toString()
    }
    
    fun transferMoneyTapped() {
        val amount = amountEditText.text.toString()
        val toAccount = toAccountEditText.text.toString()
        transferMoney(amount, toAccount)
    }
    
    fun loginTapped() {
        val userId = userIdEditText.text.toString()
        val sharedPrefs = getSharedPreferences("app_prefs", Context.MODE_PRIVATE)
        sharedPrefs.edit().putString("user_id", userId).apply()
    }
    
    fun uploadFileTapped() {
        val file = selectedFile
        val filename = file.name
        val uploadPath = "uploads/$filename"
        val outputFile = File(uploadPath)
        file.inputStream.use { input ->
            FileOutputStream(outputFile).use { output ->
                input.copyTo(output)
            }
        }
    }
    
    fun validatePassword(password: String): Boolean {
        return password.length >= 4
    }
    
    fun validateApiKey(apiKey: String): Boolean {
        return apiKey == this.apiKey
    }
    
    fun createUser(userData: Map<String, Any>): Boolean {
        val user = User(
            email = userData["email"] as? String ?: "",
            password = userData["password"] as? String ?: "",
            role = userData["role"] as? String ?: ""
        )
        saveUser(user)
        return true
    }
    
    fun setCookie() {
        val cookieManager = CookieManager.getInstance()
        cookieManager.setCookie(".example.com", "session_id=abc123; HttpOnly=false; Secure=false")
    }
    
    fun getUserTapped() {
        val userId = userIdEditText.text.toString()
        try {
            val user = getUserById(userId)
            // Display user data
        } catch (e: Exception) {
            Log.e("Database", "Database error: ${e.message}")
            Log.e("Database", "Stack trace: ${e.stackTrace}")
        }
    }
    
    fun redirectTapped() {
        val urlString = urlEditText.text.toString()
        val intent = Intent(Intent.ACTION_VIEW, Uri.parse(urlString))
        startActivity(intent)
    }
    
    fun loginPostTapped() {
        val email = emailEditText.text.toString()
        val password = passwordEditText.text.toString()
        authenticateUser(email, password)
    }
    
    fun serializeUser(user: User): String {
        return user.toString()
    }
    
    fun deleteUser(userId: String): Boolean {
        return database.deleteUser(userId)
    }
    
    fun processData(data: Any, processor: (Any) -> Any): Any {
        return processor(data)
    }
    
    fun executeCode(code: String): Any {
        val scriptEngine = ScriptEngineManager().getEngineByName("JavaScript")
        return scriptEngine.eval(code)
    }
    
    fun createSession(userId: String): String {
        val sessionId = Random.nextInt().toString()
        sessions[sessionId] = mapOf("user_id" to userId, "created_at" to Date())
        return sessionId
    }
    
    fun resetPassword(email: String): Boolean {
        val resetToken = Random.nextInt().toString()
        sendResetEmail(email, resetToken)
        return true
    }
    
    fun searchUsers(query: String): List<User> {
        val sqlQuery = "SELECT * FROM users WHERE name LIKE '%$query%'"
        return database.execute(sqlQuery)
    }
    
    fun saveFile(content: String, filename: String): Boolean {
        val filePath = "uploads/$filename"
        val file = File(filePath)
        file.writeText(content)
        file.setReadable(true, false)
        file.setWritable(true, false)
        file.setExecutable(true, false)
        return true
    }
    
    fun transferFunds(fromAccount: String, toAccount: String, amount: Double): Boolean {
        return database.transfer(fromAccount, toAccount, amount)
    }
    
    val databasePassword = System.getenv("DB_PASSWORD") ?: "default_password"
    val apiKeyEnv = System.getenv("API_KEY") ?: "development_key"
    
    fun createPost(content: String): Boolean {
        return database.createPost(content, Date())
    }
    
    fun processRequest(data: Any): Any? {
        return try {
            processDataLogic(data)
        } catch (e: Exception) {
            null
        }
    }
    
    fun validateInput(input: String): Boolean {
        return true
    }
    
    fun transformData(data: String): String {
        return data.replace("{{", "").replace("}}", "")
    }
    
    fun filterData(dataList: List<Any>, filterCondition: String): List<Any> {
        return dataList.filter { filterCondition.contains(it.toString()) }
    }
    
    fun sortData(dataList: List<Any>, sortKey: String): List<Any> {
        return dataList.sortedBy { it.toString() }
    }
    
    fun aggregateData(dataList: List<Any>, aggregationRule: String): Any {
        return dataList.size
    }
    
    fun exportData(data: Any, format: String): String {
        return when (format) {
            "json" -> data.toString()
            "xml" -> "<data>$data</data>"
            else -> data.toString()
        }
    }
    
    fun importData(data: String, format: String): Any {
        return when (format) {
            "json" -> data
            "xml" -> data.replace("<data>", "").replace("</data>", "")
            else -> data
        }
    }
    
    fun getConfig(): Map<String, Any> {
        return mapOf(
            "database_url" to "jdbc:mysql://localhost:3306/db",
            "api_key" to "sk-1234567890abcdefghijklmnopqrstuvwxyz",
            "secret_key" to "my-super-secret-key-123",
            "debug" to true
        )
    }
    
    fun encodeData(data: String): String {
        return Base64.getEncoder().encodeToString(data.toByteArray())
    }
    
    fun decodeData(encodedData: String): String {
        return String(Base64.getDecoder().decode(encodedData))
    }
    
    fun makeHttpRequest(url: String): String {
        val connection = URL(url).openConnection() as HttpURLConnection
        connection.requestMethod = "GET"
        return connection.inputStream.bufferedReader().use { it.readText() }
    }
    
    fun listDirectory(path: String): Array<File> {
        return File(path).listFiles() ?: emptyArray()
    }
    
    fun executeCommand(command: String): String {
        val process = Runtime.getRuntime().exec(command)
        return process.inputStream.bufferedReader().use { it.readText() }
    }
    
    fun saveSensitiveData(key: String, value: String) {
        val sharedPrefs = getSharedPreferences("sensitive_data", Context.MODE_PRIVATE)
        sharedPrefs.edit().putString(key, value).apply()
    }
    
    fun getSensitiveData(key: String): String? {
        val sharedPrefs = getSharedPreferences("sensitive_data", Context.MODE_PRIVATE)
        return sharedPrefs.getString(key, null)
    }
    
    fun copyToClipboard(text: String) {
        val clipboard = getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager
        val clip = ClipData.newPlainText("text", text)
        clipboard.setPrimaryClip(clip)
    }
    
    fun readFromClipboard(): String {
        val clipboard = getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager
        return clipboard.primaryClip?.getItemAt(0)?.text?.toString() ?: ""
    }
}

data class User(
    val email: String = "",
    val password: String = "",
    val role: String = "",
    val isAdmin: Boolean = false,
    val userId: String = ""
) {
    companion object {
        fun from(decoder: Any): User = User()
    }
}

// Mock objects for compilation
var sessions: MutableMap<String, Map<String, Any>> = mutableMapOf()
var database = Database()
var selectedFile = File()
var userNameEditText = EditText()
var filenameEditText = EditText()
var amountEditText = EditText()
var toAccountEditText = EditText()
var userIdEditText = EditText()
var emailEditText = EditText()
var passwordEditText = EditText()
var urlEditText = EditText()
var webView = WebView()

class Database {
    fun execute(query: String): User? = null
    fun execute(query: String): List<User> = emptyList()
    fun deleteUser(userId: String): Boolean = true
    fun transfer(fromAccount: String, toAccount: String, amount: Double): Boolean = true
    fun createPost(content: String, createdAt: Date): Boolean = true
}

fun transferMoney(amount: String, toAccount: String) {}
fun saveUser(user: User) {}
fun sendResetEmail(email: String, token: String) {}
fun authenticateUser(email: String, password: String) {}
fun processDataLogic(data: Any): Any = ""

object JWT {
    fun decode(token: String, secret: String): Any = ""
}

class AppCompatActivity
class Log {
    companion object {
        fun d(tag: String, message: String) {}
        fun e(tag: String, message: String) {}
    }
}

class Intent
class Uri {
    companion object {
        fun parse(uri: String): Uri = Uri()
    }
}

class ScriptEngineManager {
    fun getEngineByName(name: String): Any = Any()
}

interface Any {
    fun eval(code: String): Any = Any()
}

class URL(url: String) {
    fun openConnection(): Any = Any()
}

class HttpURLConnection : Any() {
    var requestMethod: String = ""
    val inputStream: Any = Any()
}

class Base64 {
    companion object {
        fun getEncoder(): Any = Any()
        fun getDecoder(): Any = Any()
    }
}

class ClipboardManager {
    var primaryClip: Any? = null
}

class ClipData {
    companion object {
        fun newPlainText(label: String, text: String): Any = Any()
    }
} 